allel/stats/roh.py

```diff
@@ -0,0 +1,231 @@
+    is_accessible: array_like, bool, shape (npositions)
+        Boolean array for each locus in contig describing whether accessible
+        or not.
+    contig_size: int
+        If is_accessible not known/not provided, allows specification of
+        total length of contig.
+    plot: bool
+        Whether to plot the ROH with heterzygosity and inaccessible regions
+    plotwinsize: int
+        Size of window in which to summarize heterozygosity for the plot
+
+
+    Returns
+    -------
+    roh: ndarray, int, shape (n_roh, 2)
+        Span windows of ROH
+    froh: float
+        Proportion of genome in ROH
+
+    """
+
+    from hmmlearn import hmm
+
+    if isinstance(phet_nonroh, float):
+        phet_nonroh = np.array((phet_nonroh,))
+
+    emission_px = np.concatenate([(phet_roh, ), phet_nonroh])
+
+    # start probability
+    start_prob = np.repeat(1/emission_px.size, emission_px.size)
+
+    # transition between underlying states
+    transitions = _derive_transition_mx(transition, emission_px.size)
+
+    # probability of inaccessible
+    if is_accessible is None:
+        assert contig_size is not None, \
+            "If accessibility not provided must specify size of contig"
+        p_accessible = 1.0
+    else:
+        p_accessible = is_accessible.mean()
+        assert contig_size is None, "Contig size only specified when " \
+                                    "is_accessible is not provided"
+        contig_size = is_accessible.size
+
+    emission_mx = _derive_emission_mx(emission_px, p_accessible)
+
+    # initialize HMM
+    roh_hmm = hmm.MultinomialHMM(n_components=emission_px.size)
+
+    roh_hmm.n_symbols_ = 3
+    roh_hmm.startprob_ = start_prob
+    roh_hmm.transmat_ = transitions
+    roh_hmm.emissionprob_ = emission_mx
+
+    gsub = g.take([ix], axis=1)
+    hz = gsub.count_het(axis=1).astype("bool")
+
+    pred, obs = _predict_roh_state(roh_hmm, hz, pos, is_accessible, contig_size)
+    homozygous_windows = _get_state_windows(pred, state=0)
+
+    # filter by roh size
+    roh_sizes = np.diff(homozygous_windows, axis=1).flatten()
+    roh = np.compress(roh_sizes >= min_roh, homozygous_windows, axis=0)
+    froh = np.diff(roh, axis=1).sum() / contig_size
+
+    if plot:
+        import matplotlib.pyplot as plt
+        from allel.stats import windowed_statistic
+
+        window = np.arange(0, is_accessible.size, plotwinsize)
+        windows = np.vstack([window[:-1], window[1:]]).T
+
+        count_hz, win, count = windowed_statistic(pos=pos, values=hz,
+                                                  statistic=np.sum,
+                                                  windows=windows)
+
+        if is_accessible is not None:
+
+            # remove the last val as not a complete window
+            count_acc = np.add.reduceat(
+                is_accessible, window)[:-1]
+
+            # if accessible bases in a window < 1000, don't plot.
+            # b/c Leads to spiky data
```

Comment:
For the moment I have taken the plotting code out of the PR. The plotting code needs some of the intermediate variables from the roh function: pos, is_het etc, so I can't think of a good way for it to operate on the output of the main roh function.